### PR TITLE
Add Adapters to ALBERT

### PR DESCRIPTION
**Pull request by lenglaender**
This PR adds adapters to the ALBERT model.

All tests run through except some class conversation tests. I think the error must be in the `STATIC_TO_FLEX_HEAD_MAP` in `head_utils.py`, but I am not 100% sure. Can you take a look at this, @calpt or @hSterz ?

### CODE REVIEW

File: src/transformers/adapters/mixins/albert.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        self.attention_adapters._init_adapter_modules()
+
+
+class AlbertModelAdaptersMixin(EmbeddingAdaptersMixin, InvertibleAdaptersMixin, ModelAdaptersMixin):
+    """Adds adapters to the AlbertModel module."""
+
+    def iter_layers(self) -> Iterable[Tuple[int, nn.Module]]:
+        for i, albertLayerGroup in enumerate(self.encoder.albert_layer_groups):
+            for albertLayer in albertLayerGroup.albert_layers:
+                yield i, albertLayer
```
Comment (calpt) [inline src/transformers/adapters/mixins/albert.py:29]: I'm not sure about the exact logic of Albert's layer sharing, but couldn't treating the layer group id here as layer id for each adapter be problematic for certain config keys (e.g. reduction factor per layer, leave_out)?

File: tests_adapters/test_adapter_embeddings.py [hunk 2]
```diff
@@ -21,34 +21,34 @@ def test_load_embeddings(self):
 
         self.assertEqual(model.active_embeddings, "test")
 
-    def test_add_embeddings(self):
+    def test_add_embeddings(self, embedding_dim=None):
         model = self.get_model()
         tokenizer = AutoTokenizer.from_pretrained("tests_adapters/fixtures/SiBERT")
-        model.add_embeddings("test", tokenizer)
+        model.add_embeddings("test", tokenizer, embedding_dim=embedding_dim)
```
Comment (calpt) [inline tests_adapters/test_adapter_embeddings.py:27]: Here and other places with this change: Shouldn't we rather try to figure out `embedding_dim` automatically in `add_embeddings()` as `embedding_size` is part of the model config?

**Review by calpt — COMMENTED** [review]
Looks pretty good to me overall 👍

File: src/transformers/adapters/mixins/albert.py (hunk 1 above)
Comment (lenglaender) [reply src/transformers/adapters/mixins/albert.py]: You are right - should be fixed by acefa59c3b3ebbf65eb92f140ca0fed492e4ed18

**Comment by lenglaender** [comment]
Thanks for the review @calpt. Everything you noticed should be fixed now

File: src/transformers/adapters/mixins/albert.py (hunk 1 above)
Comment (calpt) [reply src/transformers/adapters/mixins/albert.py]: Thanks for fixing.

I think we should additionally add a note in the docs to clarify how the adapter implementation in Albert is different from others (i.e. that layers + adapters are shared and using layer ids in the adapter config will not affect the actual Albert layers, but the hidden layer groups + inner groups). I did a similar thing for CLIP here: https://raw.githubusercontent.com/adapter-hub/adapter-transformers/master/adapter_docs/classes/models/clip.rst

File: src/transformers/adapters/model_mixin.py [hunk 3]
```diff
@@ -188,7 +193,7 @@ def load_embeddings(self, path: str, name: str):
         self.set_active_embeddings(name)
         return tokenizer
 
-    def add_embeddings(self, name, tokenizer, reference_embedding=None, reference_tokenizer=None, embedding_dim=None):
```
Comment (calpt) [inline src/transformers/adapters/model_mixin.py:191]: I'm not sure if the `embedding_dim` parameter can be removed without issues (as it was available before this PR). If we want to remove it, also remove the method docs of this param below.

**Review by calpt — APPROVED** [review]
Feel free to squash & merge once remaining two comments are resolved.

File: src/transformers/adapters/model_mixin.py (hunk 3 above)
Comment (lenglaender) [reply src/transformers/adapters/model_mixin.py]: You're right, I added the parameter again

File: src/transformers/adapters/mixins/albert.py (hunk 1 above)
Comment (lenglaender) [reply src/transformers/adapters/mixins/albert.py]: Added a note to the docs in 4d2fb06 :+1:

**Pull request merged**